File: 0046-Permutations/0046.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) <= 1:
            return [nums]
        
        out = []
        perms = self.permute(nums[1:])
        for perm in perms:
            for i in range(0, len(perm)+1):
                p = perm[:i] + [nums[0]] + perm[i:]
                out.append(p)
        
        return out


if __name__ == '__main__':
    nums = [1,2,3]
    print(Solution().permute(nums))

File: src/0046-Permutations/0046.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution
{
public:
    vector<vector<int>> permute(vector<int>& nums)
    {
        if (nums.size() <= 1) return vector<vector<int>>{nums};
        vector<vector<int>> out;
        auto tmpNums = vector<int>(nums.begin() + 1, nums.end());
        auto perms = permute(tmpNums);
        for (auto& perm : perms)
        {
            for (unsigned int i = 0; i <= perm.size(); ++i)
            {
            	auto tmp_perm = perm;
            	tmp_perm.insert(tmp_perm.begin() + i, nums[0]);
                out.push_back(tmp_perm);
            }
        }
        return out;
    }
};